packages/lodestar/src/chain/chain.ts

```diff
@@ -193,6 +195,74 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
+        if (this.config.types.Root.equals(root, blockRoot)) {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Restore state cache and forkchoice from last finalized state.
+   */
+  private async restoreHeadState(): Promise<void> {
+    this.logger.profile("restoreHeadState");
+    const lastKnownState: TreeBacked<BeaconState> = await this.db.stateArchive.lastValue();
+    this.db.stateCache.add(lastKnownState);
+    // the block respective to finalized epoch is still in block db
+    const allBlocks = await this.db.block.values();
+    if (!allBlocks || allBlocks.length === 0) {
+      return;
+    }
+    this.logger.info(`Found ${allBlocks.length} blocks in database`);
+    if (allBlocks.length === 1) {
+      // start from scratch
+      const block = allBlocks[0];
+      const blockHash = this.config.types.BeaconBlock.hashTreeRoot(block.message);
+      if (this.config.types.Root.equals(blockHash, this.forkChoice.headBlockRoot())) {
+        this.logger.info("Chain is up to date, no need to restore");
+        return;
+      }
+    }
+    const lastKnownStateRoot = this.config.types.BeaconState.hashTreeRoot(lastKnownState);
+    const sortedBlocks = sortBlocks(allBlocks);
+    const isCheckpointNotGenesis = lastKnownState.slot > GENESIS_SLOT;
+    const finalizedBlock = sortedBlocks.find(block => {
+      return (block.message.slot === lastKnownState.slot) &&
+      // at genesis the genesis block's state root is not equal to genesis state root
+      (isCheckpointNotGenesis? this.config.types.Root.equals(block.message.stateRoot, lastKnownStateRoot) : true);
+    });
+    if (!finalizedBlock) {
+      this.logger.error(`Cannot find block for finalized state at slot ${lastKnownState.slot}`);
+    } else {
+      this.logger.info(`Found finalized block at slot ${finalizedBlock.message.slot}`);
+    }
+    const blockCheckpoint = {
+      root: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      epoch: computeEpochAtSlot(this.config, finalizedBlock.message.slot)
+    };
+    this.forkChoice.addBlock({
+      slot: finalizedBlock.message.slot,
+      blockRoot: this.config.types.BeaconBlock.hashTreeRoot(finalizedBlock.message),
+      stateRoot: finalizedBlock.message.stateRoot.valueOf() as Uint8Array,
+      parentRoot: finalizedBlock.message.parentRoot.valueOf() as Uint8Array,
+      justifiedCheckpoint: (isCheckpointNotGenesis)? lastKnownState.currentJustifiedCheckpoint : blockCheckpoint,
+      finalizedCheckpoint: (isCheckpointNotGenesis)? lastKnownState.finalizedCheckpoint : blockCheckpoint
+    });
+
+    for (const block of sortedBlocks) {
+      await this.receiveBlock(block, true);
+    }
+    const lastBlock = sortedBlocks[sortedBlocks.length - 1];
+    await this.waitForBlockProcessed(this.config.types.BeaconBlock.hashTreeRoot(lastBlock.message));
+    this.logger.important(`Finish restoring chain head from ${allBlocks.length} blocks`);
+    this.logger.profile("restoreHeadState");
```

Comment:
It only prints `restoreHeadState` without duration. Any ideas @mpetrunic ?

Reply:
My best guess is because of our custom log formatting. It ignores metadata.
You should log `info` object here https://github.com/ChainSafe/lodestar/blob/master/packages/lodestar-utils/src/logger/winston.ts#L32

I really wish we fix that so whatever metadata we sent there is logged like "<propertyName1>=<value1>, <propertyName2>=<value2>" at the end of log message

Reply:
:100: :100: i've been frustrated w our logger setup. I'd love to see more data exposed via a context object. (or more generally thinking of our logs as json-serializable objects) IMO we should at minimum fix up winston, possibly revisit other alternatives like pino and https://github.com/pinojs/pino-pretty . (bonus: it plays nicely w/ fastify)

Reply:
Ill fix logger in next pr so we can log context passed along with message